File: VB/Controls/MyGridControl.vb
Imports System.Windows
Imports System.Collections
Imports System.Windows.Controls
Imports DevExpress.Xpf.Grid
Imports DXGridSample.Extensions
Imports System.Collections.Specialized

Namespace DXGridSample.Controls

    Public Class MyGridControl
        Inherits GridControl

        Public Shared ReadOnly BandTemplateProperty As DependencyProperty = DependencyProperty.Register("BandTemplate", GetType(DataTemplate), GetType(MyGridControl), New PropertyMetadata(CType(Nothing, PropertyChangedCallback)))

        Public Property BandTemplate As DataTemplate
            Get
                Return CType(GetValue(BandTemplateProperty), DataTemplate)
            End Get

            Set(ByVal value As DataTemplate)
                SetValue(BandTemplateProperty, value)
            End Set
        End Property

        Public Shared ReadOnly BandsSourceProperty As DependencyProperty = DependencyProperty.Register("BandsSource", GetType(IList), GetType(MyGridControl), New PropertyMetadata(Nothing, AddressOf DXGridSample.Controls.MyGridControl.OnBandedSourcePropertyChanged))

        Public Property BandsSource As IList
            Get
                Return CType(GetValue(BandsSourceProperty), IList)
            End Get

            Set(ByVal value As IList)
                SetValue(BandsSourceProperty, value)
            End Set
        End Property

        Private Shared Sub OnBandedSourcePropertyChanged(ByVal d As DependencyObject, ByVal e As DependencyPropertyChangedEventArgs)
            CType(d, MyGridControl).PopulateBands()
            Dim myDelegate As NotifyCollectionChangedEventHandler = Sub(ByVal sender, ByVal par) CType(d, MyGridControl).PopulateBands()
            If e.OldValue IsNot Nothing AndAlso TypeOf e.OldValue Is INotifyCollectionChanged Then
                RemoveHandler CType(e.NewValue, INotifyCollectionChanged).CollectionChanged, myDelegate
            End If

            If e.NewValue IsNot Nothing AndAlso TypeOf e.NewValue Is INotifyCollectionChanged Then
                AddHandler CType(e.NewValue, INotifyCollectionChanged).CollectionChanged, myDelegate
            End If
        End Sub

        Private Sub PopulateBands()
            Bands.Clear()
            For Each b In BandsSource
                Dim cc As ContentControl = TryCast(BandTemplate.LoadContent(), ContentControl)
                If cc Is Nothing Then Continue For
                Dim band As GridControlBand = TryCast(cc.Content, GridControlBand)
                cc.Content = Nothing
                If band Is Nothing Then Continue For
                band.DataContext = b
                DependencyObjectHelper.SetDataContext(band, band.DataContext)
                Bands.Add(band)
            Next
        End Sub
    End Class
End Namespace

File: CS/Controls/MyGridControl.cs
using System.Windows;
using System.Collections;
using System.Windows.Controls;
using DevExpress.Xpf.Grid;
using DXGridSample.Extensions;
using System.Collections.Specialized;
using System;

namespace DXGridSample.Controls {
    public class MyGridControl : GridControl {
        public static readonly DependencyProperty BandTemplateProperty =
            DependencyProperty.Register("BandTemplate", typeof(DataTemplate), typeof(MyGridControl), new PropertyMetadata(null));
        public DataTemplate BandTemplate {
            get { return (DataTemplate)GetValue(BandTemplateProperty); }
            set { SetValue(BandTemplateProperty, value); }
        }
        public static readonly DependencyProperty BandsSourceProperty =
            DependencyProperty.Register("BandsSource", typeof(IList), typeof(MyGridControl), new PropertyMetadata(null, OnBandedSourcePropertyChanged));
        public IList BandsSource {
            get { return (IList)GetValue(BandsSourceProperty); }
            set { SetValue(BandsSourceProperty, value); }
        }
        private static void OnBandedSourcePropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e) {
            ((MyGridControl)d).PopulateBands();

            NotifyCollectionChangedEventHandler myDelegate = delegate (object sender, NotifyCollectionChangedEventArgs par) { ((MyGridControl)d).PopulateBands(); };

            if (e.OldValue != null && e.OldValue is INotifyCollectionChanged) {
                ((INotifyCollectionChanged)e.NewValue).CollectionChanged -= myDelegate;
            }

            if (e.NewValue != null && e.NewValue is INotifyCollectionChanged) {
                ((INotifyCollectionChanged)e.NewValue).CollectionChanged += myDelegate;
            }

        }

        private void PopulateBands() {
            Bands.Clear();
            foreach (var b in BandsSource) {
                ContentControl cc = BandTemplate.LoadContent() as ContentControl;
                if (cc == null)
                    continue;
                GridControlBand band = cc.Content as GridControlBand;
                cc.Content = null;
                if (band == null)
                    continue;
                band.DataContext = b;
                DependencyObjectHelper.SetDataContext(band, band.DataContext);                
                Bands.Add(band);
            }
        }
    }
}